src/Service/ProductPositionsUpdater.php

```diff
@@ -0,0 +1,117 @@
+        private RouterInterface $router,
+        private EventDispatcherInterface $eventDispatcher,
+        private TranslatorInterface $translator,
+    )
+    {
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    public function process(Request $request): string
+    {
+        $this->initialize($request);
+
+        if ($this->identifiersPositionsMap === []) {
+            $this->setSessionError('threebrs.ui.sortingPlugin.chooseTaxon');
+        }
+
+        $productTaxonCollection = $this->entityManager
+            ->getRepository(ProductTaxonInterface::class)
+            ->findBy(['id' => array_keys($this->identifiersPositionsMap)]);
+
+        if (count($productTaxonCollection) !== count($this->identifiersPositionsMap)) {
+            throw new ProductPositionsUpdaterException('Required identifiers set is invalid.');
+        }
+
+        $taxon = null;
+
+        foreach ($productTaxonCollection as $productTaxon) {
+            assert($productTaxon instanceof ProductTaxonInterface);
+            $taxon ??= $productTaxon->getTaxon();
+            $productTaxon->setPosition($this->identifiersPositionsMap[$productTaxon->getId()] ?? 0);
+        }
+
+        if ($taxon === null) {
+            $this->setSessionError('threebrs.ui.sortingPlugin.noProductMessage');
+
+            return $this->router->generate('threebrs_admin_sorting_index');
+        }
+
+        $this->entityManager->flush();
+        $this->setSessionSuccess('threebrs.ui.sortingPlugin.successMessage');
+        $this->eventDispatcher->dispatch(new GenericEvent($taxon), 'threebrs-sorting-products-after-persist');
+
+        return $this->router->generate('threebrs_admin_sorting_products', ['taxonId' => $taxon->getId()]);
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    private function prepareSessionMessage(string $key): string
+    {
+        if ($this->sessionBag === null) {
+            throw new ProductPositionsUpdaterException('Session\'s state is corrupted.');
+        }
+
+        return $this->translator->trans($key);
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    private function setSessionError(string $messageKey): void
+    {
+        $this->sessionBag->add(
+            'error',
+            $this->prepareSessionMessage($messageKey)
+        );
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    private function setSessionSuccess(string $messageKey): void
+    {
+        $this->sessionBag->add(
+            'success',
+            $this->prepareSessionMessage($messageKey),
+        );
+    }
+
+    private function initialize(Request $request): void
+    {
+        $sessionBag = $request->getSession()->getBag('flashes');
+        assert($sessionBag instanceof FlashBagInterface);
+        $this->sessionBag = $sessionBag;
+
+        $this->identifiersPositionsMap = array_flip(array_values($request->get('id', [])));
```

Comment:
Můžeš si to připravit rovnou v metodě `process`, když má metoda jen jeden řádek a používá se jen na jednom místě

Reply:
Jo proč ne. Týká se to přímo dat ke zpracování, takže to asi patří víc do `process` než do inicializace (přípravy prostředí).

Přesunul jsem to přímo do `process`